json-form-react/utilities/JsonFormValidator.js

```diff
@@ -110,4 +120,115 @@ export default class JsonFormValidator {
     return null;
   }
 
+  _validateEmail(value) {
+
+    const validRegex = /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/;
+
+    if (!value.match(validRegex)) {
+
+      return 'invaid email type';
+    }
+    return null;
+  }
+
+
+  _validateDate(value) {
+
+    let isValidDate = Date.parse(value);
+
+    if (isNaN(isValidDate)) {
+      return 'Invalid Date Format';
+    }
+    return null;
+  }
+
+
+  _validateLetterOnly(value) {
+
+    const regEx = /^[A-Za-z]+$/;
+    if (!value.match(regEx)) {
+
+      return 'Please Enter Alphabetic Letters Only';
+    }
+    return null;
+  }
+
+  _validateNumbersOnly(value) {
+
+    const numbers = /^[0-9]+$/;
+    if (!value.match(numbers)) {
+      return 'Please Enter Only Number';
+    }
+
+    return null;
+  }
+
+  _validateDate1(value) {
+    var dateformat = /^(0?[1-9]|[12][0-9]|3[01])[\/\-](0?[1-9]|1[012])[\/\-]\d{4}$/;
+
+    if (value.match(dateformat)) {
+
+      var opera1 = value.split('/');
+      var opera2 = value.split('-');
+      lopera1 = opera1.length;
+      lopera2 = opera2.length;
+
+      if (lopera1 > 1) {
+        var pdate = value.split('/');
+      }
+      else if (lopera2 > 1) {
+        var pdate = value.split('-');
+      }
+      var dd = parseInt(pdate[0]);
+      var mm = parseInt(pdate[1]);
+      var yy = parseInt(pdate[2]);
+
+      var ListofDays = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
+      if (mm == 1 || mm > 2) {
+        if (dd > ListofDays[mm - 1]) {
+
+          return 'Invalid date format!';
+        }
+      }
+      if (mm == 2) {
+        var lyear = false;
+        if ((!(yy % 4) && yy % 100) || !(yy % 400)) {
+          lyear = true;
+        }
+        if ((lyear == false) && (dd >= 29)) {
+
+          return 'Invalid date format!';
+        }
+        if ((lyear == true) && (dd > 29)) {
+
+          return 'Invalid date format!';
+        }
+      }
+    }
+    else {
+
+      return 'Invalid date format!';
+    }
+  }
+
+  _validatePhoneNumber(value) {
+
+    if (value.length < 10) {
```

Comment:
@hariidev We need to be a bit more careful with this. I can enter a Sri Lankan phone number as `0777995648` or `777995648` or also `777 99 56 48` or `77 7 995 648` and all these will be valid inputs. Also you haven't checked if the entered value is letters or numbers, only the length.

So let's make these modifications:
- filter out any spaces (even multiple spaces), brackets `( )` and dashes `-`
- filter out the leading zero if there is one
- then check if there is nine digits and it is fully numeric, no other characters

Reply:
@cie6868  if we give input type as NUMBER then we cant  enter special characters, strings and spaces.

Reply:
@hariidev Ok this makes sense for front-end valdation, but remember the validation also happens on the server. We don't know if the server will receive some broken data from another place with a normal text field. So best to be thorough in the validation.